cli/lsp/semantic_tokens.rs

```diff
@@ -247,6 +248,54 @@ impl SemanticTokensBuilder {
   }
 }
 
+pub fn tokens_within_range(
+  tokens: &SemanticTokens,
+  range: lsp::Range,
+) -> SemanticTokens {
+  let mut line = 0;
+  let mut character = 0;
+
+  let mut start_line = 0;
+  let mut start_char = 0;
+  let mut keep_start_idx = tokens.data.len();
+  let mut keep_end_idx = keep_start_idx;
+  for (i, token) in tokens.data.iter().enumerate() {
+    if token.delta_line != 0 {
+      character = 0;
+    }
+    line += token.delta_line;
+    character += token.delta_start;
+    let token_start = lsp::Position::new(line, character);
+    if i < keep_start_idx && token_start >= range.start {
+      keep_start_idx = i;
+      start_line = line;
+      start_char = character;
+    }
+    if token_start > range.end {
+      keep_end_idx = i;
```

Comment:
What would happen if the range ended at the last token? Seems like this would panic if never set?

Reply:
It would finish the for loop and `keep_end_idx` would be `tokens.data.len()` (its initial value), so we would just slice to the end of the vec

Reply:
Ah, I missed that..